docs_src/design/adr/security/0027-authentication.md

```diff
@@ -0,0 +1,257 @@
+URL rewriting is used to concentrate microservices
+on a single HTTP-accessible port.
+
+Behind the proxy, it is not possible to verify Kong
+as the origin of local network traffic because mutual-auth TLS
+is not supported in the open source version of Kong.
+Although the Kong JWT plugin will set request headers
+on the backend request that identify the caller,
+there is no mechanism by which Kong can prove to a
+backend service that it was the component that
+performed the authentication step.
+Even though the original JWT passes through the proxy,
+the Kong authentication plugins do not expose
+token introspection endpoints that the backend service
+could use to check token validity independently.
+
+The consequence of having an API gateway that performs
+all microservice authentication is that communication
+between EdgeX microservices running behind the API gateway
+are not authenticated in any way.
+EdgeX microservices are unable to distinguish
+malicious traffic that has evaded the API gateway
+from legitimate microservice traffic.
+
+## Proposed Design
+
+This ADR proposes an implementation of the
+[Microservice Authentication UCR](../../ucr/Microservice-Authentication.md)
+that uses a token-based authentication mechanism.
+
+This ADR proposes to relieve the Kong API gateway of its
+[JWT](https://www.rfc-editor.org/rfc/rfc7519) management responsibility,
+and instead use Hashicorp Vault for this purpose,
+which is already used as EdgeX's secret store.
+In turn, EdgeX can use a simpler reverse proxy, NGINX,
+resulting in a total savings of 150MB of memory
+and 300MB of storage (in the docker implementation).
+This change requires minimal modification of existing
+clients written to perform JWT-based authentication at the Kong gateway:
+they simply use a Vault-issued JWT
+instead of a Kong-issued JWT or a self-issued JWT.
+
+This ADR proposes a layered authentication scheme,
+with the reverse proxy performing an initial check for all external requests,
+and EdgeX services themselves authenticating all internal and external requests.
+The layered approach is required,
+as some EdgeX endpoints such as `/api/v2/ping`
+must be anonymous for health-checking purposes.
+
+EdgeX microservices shall consult Vault to confirm JWT validity,
+and an NGINX reverse proxy shall use the
+[ngx_http_auth_request_module](http://nginx.org/en/docs/http/ngx_http_auth_request_module.html)
+to delegate confirmation of JWT validity to Vault,
+as a mitigation for microservice-level implementation errors.
+TLS termination at the reverse proxy shall be enabled by default
+so as to be consistent with
+[ADR 0015 - Encryption between microservices](./0015-in-cluster-tls.md).
+
+Behind the proxy, there are two major changes:
+
+1. Every EdgeX service, when security is enabled,
+   requires a JWT be passed as part of the HTTP request
+   that is validated using Vault's token introspection endpoint,
+   or manually validated based on published signature keys.
```

Comment:
Not mandating either one, but I am implementing the first.

Let's get it implemented and benchmark it to see if any potential scaling issues are real or imagined.  To do otherwise is premature optimization.

Reply:
Well-known URL: https://developer.hashicorp.com/vault/api-docs/secret/identity/tokens#read-well-known-configurations
Keys URL specifically: https://developer.hashicorp.com/vault/api-docs/secret/identity/tokens#read-active-public-keys